### PR TITLE
ARROW-10106: [FlightRPC][Java] Expose onIsReady() callback

### PR DESCRIPTION
**Pull request by jduo**
Expose onIsReady() callback on OutboundStreamListener

- Add callback to run when the client is ready to receive new data during getStream.
- This removes the need for FlightProducers to implement polling code on isReady.

### CODE REVIEW

**Comment by jduo** [comment]
FYI @lidavidm

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/OutboundStreamListener.java [hunk 1]
```diff
@@ -36,6 +36,18 @@
    */
   boolean isReady();
 
+  /**
+   * Set a callback for when the client cancels is ready for new calls to putNext(), i.e. {@link #isReady()} ()}
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/OutboundStreamListener.java:40]:
```suggestion
   * Set a callback for when the listener is ready for new calls to putNext(), i.e. {@link #isReady()} ()}
```

**Review by lidavidm — COMMENTED** [review]
LGTM, one minor typo - and is it worth adding a test? (Though this is a very thin wrapper around gRPC.)

**Comment by jduo** [comment]
> LGTM, one minor typo - and is it worth adding a test? (Though this is a very thin wrapper around gRPC.)

Thanks. I'm looking to repurpose some existing tests that loop on isReady.

**Comment by jduo** [comment]
> > LGTM, one minor typo - and is it worth adding a test? (Though this is a very thin wrapper around gRPC.)
> 
> Thanks. I'm looking to repurpose some existing tests that loop on isReady.

- I've created a BackpressureStrategy interface and a simple callback-based implementation.
- I've used this to repurpose the tests in TestBackPressure to test both polling (existing) and callback-based solutions.
- New tests in TestBackpressure are marked Ignored since the tests they were based on were also ignored due to flakiness. I ran them locally though and they passed with similar performance.
- I needed to change getStream() to run in a background thread. Currently it gets run in the onHalfClosed callback handler in gRPC. In the test implementations, getStream() is a synchronous implementation which prevented the new callback handler from running. In practice, I'd expect real world getStream() implementations to be written asynchronously though, so perhaps we should change our test implementations of getStream() to be asynchronous instead.

This line of code eventually calls doGet, which must complete before onReady is called (thus, if doGet blocks until a notification from onReady, it never completes. But polling on isReady does): https://github.com/grpc/grpc-java/blob/0b6f29371bd96614fdbdcd3638d4bb6312258da3/stub/src/main/java/io/grpc/stub/ServerCalls.java#L182

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        if (listener.isReady()) {
+          return WaitResult.READY;
+        } else if (listener.isCancelled()) {
+          return WaitResult.CANCELLED;
+        } else if (System.currentTimeMillis() > startTime + timeout) {
+          return WaitResult.TIMEOUT;
+        }
+        throw new RuntimeException("Invalid state when waiting for listener.");
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java:103]: According to the gRPC docs, this isn't an invalid state - gRPC can wake you up spuriously.

https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/CallStreamObserver.html#setOnReadyHandler-java.lang.Runnable-

Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java:103]: I'd say we can probably just recurse into waitForListener or wrap the method body in a loop.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java [hunk 3]
```diff
@@ -147,11 +149,23 @@ private void onCancel() {
       }
     }
 
+    private void onReady() {
+      logger.debug("Stream is ready for new messages.");
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java:153]: This is going to spam logs a lot - best not to add it unless it's really valuable.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/OutboundStreamListener.java [hunk 4]
```diff
@@ -36,6 +36,18 @@
    */
   boolean isReady();
 
+  /**
+   * Set a callback for when the listener is ready for new calls to putNext(), i.e. {@link #isReady()} ()}
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/OutboundStreamListener.java:40]: The formatting here and below seems off.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBackPressure.java [hunk 5]
```diff
@@ -158,4 +189,64 @@ private static void consume(FlightStream stream, int batches) {
...
+          sleepTime.addAndGet(1L);
+        } catch (InterruptedException ignore) {
+        }
+      }
+      return WaitResult.READY;
+    }
+  }
+
+  /**
+   * Implementation of a backpressure strategy that polls on uses callbacks to detect changes in client readiness state
```
Comment (lidavidm) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBackPressure.java:232]:
```suggestion
   * Implementation of a backpressure strategy that uses callbacks to detect changes in client readiness state
```

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java [hunk 6]
```diff
@@ -88,12 +87,15 @@ public void listFlights(Flight.Criteria criteria, StreamObserver<Flight.FlightIn
...
     final ServerCallStreamObserver<ArrowMessage> responseObserver =
         (ServerCallStreamObserver<ArrowMessage>) responseObserverSimple;
+
     final GetListener listener = new GetListener(responseObserver, this::handleExceptionWithMiddleware);
-    try {
-      producer.getStream(makeContext(responseObserver), new Ticket(ticket), listener);
-    } catch (Exception ex) {
-      listener.error(ex);
-    }
+    executors.submit(() -> {
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java:92]: I'd rather we not use the executor here unconditionally. A fully asynchronous implementation would set the callbacks and not need to execute in a separate thread at all; an implementation that wants to appear blocking but use the callbacks (as with BackpressureStrategy) should manage its own thread pool. We've had production issues when it turned out that the internal executor in FlightService was queueing requests without bound, defeating the rate limit/concurrency controls that had been placed on the gRPC executor.

Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java:92]: I think you'll have to change setOnReadyHandler below to call responseObserver.setOnReadyHandler itself

**Review by lidavidm — COMMENTED** [review]
BackpressureStrategy looks convenient, but I'd rather we not change getStream to silently execute requests on a background executor. (DoPut and DoExchange have to do so in order to present a blocking interface to consumers, but DoGet doesn't need to.)

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java (hunk 6 above)
Comment (jduo) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java:92]: Do you mean within GetListener? We register #onReady() on the responseObserver during GetListener's constructor, which will delegate to the onReadyHandler if it's been set.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java (hunk 2 above)
Comment (jduo) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java:116]: I was looking at this and I think this gets covered by the upper loop -- if we ran the onReadyCallback, we'd wake out of the Object.wait() call, but then we would check if it is really ready as part of the loop condition, and if not, go back into the wait and wait for the callback again. (Pretty much the same as a standard use of wait/notify() really).

Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java:117]: Ah yes, you're right. Slight nit then, maybe IllegalStateException is more appropriate here?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java (hunk 6 above)
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightService.java]: Sorry, yes, I meant within GetListener, but I suppose setting the handler unconditionally doesn't matter to gRPC - it'll never get called in the blocking case.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java (hunk 2 above)
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/BackpressureStrategy.java:117]: Ah yes, you're right.

**Review by lidavidm — APPROVED** [review]
LGTM, thanks for implementing this!

**Pull request closed without merging**